java-checks/src/main/java/org/sonar/java/checks/RightCurlyBraceStartLineCheck.java

```diff
@@ -41,43 +64,186 @@
+      lastToken = tree.closeBraceToken();
+    }
+
+    @Override
+    public void visitEmptyStatement(EmptyStatementTree tree) {
+      lastToken = tree.semicolonToken();
+    }
+
+    @Override
+    public void visitExpressionStatement(ExpressionStatementTree tree) {
+      lastToken = tree.semicolonToken();
+    }
+
+    @Override
+    public void visitIfStatement(IfStatementTree tree) {
+      StatementTree elseStatement = tree.elseStatement();
+      if (elseStatement != null) {
+        scan(elseStatement);
+      } else {
+        scan(tree.thenStatement());
       }
+    }
+
+    @Override
+    public void visitAssertStatement(AssertStatementTree tree) {
+      lastToken = tree.semicolonToken();
+    }
+
+    @Override
+    public void visitSwitchStatement(SwitchStatementTree tree) {
+      lastToken = tree.closeBraceToken();
+    }
+
+    @Override
+    public void visitWhileStatement(WhileStatementTree tree) {
+      scan(tree.statement());
+    }
+
+    @Override
+    public void visitDoWhileStatement(DoWhileStatementTree tree) {
+      lastToken = tree.semicolonToken();
+    }
+
+    @Override
+    public void visitForStatement(ForStatementTree tree) {
+      scan(tree.statement());
+    }
 
-      while (result != null && result.getLastChild() != null) {
-        result = result.getLastChild();
+    @Override
+    public void visitForEachStatement(ForEachStatement tree) {
+      scan(tree.statement());
+    }
+
+    @Override
+    public void visitBreakStatement(BreakStatementTree tree) {
+      lastToken = tree.semicolonToken();
+    }
+
+    @Override
+    public void visitContinueStatement(ContinueStatementTree tree) {
+      lastToken = tree.semicolonToken();
+    }
+
+    @Override
+    public void visitReturnStatement(ReturnStatementTree tree) {
+      lastToken = tree.semicolonToken();
+    }
+
+    @Override
+    public void visitThrowStatement(ThrowStatementTree tree) {
+      lastToken = tree.semicolonToken();
+    }
+
+    @Override
+    public void visitSynchronizedStatement(SynchronizedStatementTree tree) {
+      scan(tree.block());
+    }
+
+    @Override
+    public void visitTryStatement(TryStatementTree tree) {
+      BlockTree finallyBlock = tree.finallyBlock();
+      if (finallyBlock != null) {
+        scan(finallyBlock);
+      } else {
+        List<CatchTree> catches = tree.catches();
+        if (catches.isEmpty()) {
+          // with try-with-resources, catches and block are optional (JLS8 14.20.3.1)
+          scan(tree.block());
+        } else {
+          // scan only the last catch
+          scan(catches.get(catches.size() - 1));
+        }
       }
     }
 
-    return result == null ? null : result.getToken();
-  }
+    @Override
+    public void visitCatch(CatchTree tree) {
+      scan(tree.block());
+    }
+
+    @Override
+    public void visitMethod(MethodTree tree) {
+      SyntaxToken semicolonToken = tree.semicolonToken();
+      if (semicolonToken != null) {
```

Comment:
![New](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/new.png)![CRITICAL](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-critical.png)Redundant nullcheck of semicolonToken, which is known to be non-null in org.sonar.java.checks.RightCurlyBraceStartLineCheck$SyntaxTokenFinder.visitMethod(MethodTree)[![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://nemo.sonarqube.org/coding_rules#rule_key=findbugs%3ARCN_REDUNDANT_NULLCHECK_OF_NONNULL_VALUE)